app/controllers/needs_controller.rb

```diff
@@ -4,7 +4,15 @@ class NeedsController < ApplicationController
   def index
     authorize Need
 
-    @needs = policy_scope(Need).includes(:shifts).current.order(:start_at)
+    @needs = policy_scope(Need).includes(:shifts).order(:start_at)
+
+    if current_user.role == 'admin' && params[:date]
+      @date = Date.parse(params[:date])
+      @needs = @needs.on_date(@date)
+    else
+      @date = nil
+      @needs = @needs.current
+    end
```

Comment:
If the `date` param was an invalid string (for example, `abc`), `Date.parse` will blow up with an `ArgumentError`. What if we did something like this?

```ruby
def index
  authorize Need

  needs_scope = policy_scope(Need).includes(:shifts).order(:start_at)

  @date = safe_parse_date(params[:date])
  @needs = @date && current_user.admin? ? needs_scope.on_date(@date) : needs_scope.current
end

private

def safe_parse_date(date)
  Date.parse(date)
rescue ArgumentError, TypeError
end
```

Reply:
Could also write `safe_parse_date` like this:


```ruby
def safe_parse_date(date)
  date && Date.parse(date)
rescue ArgumentError
end
```

It makes it a bit more clear what is going on.

Reply:
While true, there should be no legitimate reason that a user would be manually inputting anything other than the date format from the calendar there. I think raising an exception is probably an acceptable outcome of that situation, but can change this if you prefer.

Reply:
Happy to leave it to your discretion. Please do use current_user.admin? here as well, though.